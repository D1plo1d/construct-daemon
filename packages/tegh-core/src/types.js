/* config */
export Config, { MockConfig } from './config/types/Config'
export * as AxisTypeEnum from './config/types/AxisTypeEnum'

/* jobQueue */
export * as JobHistoryTypeEnum from './jobQueue/types/JobHistoryTypeEnum'

export JobHistoryEvent from './jobQueue/types/JobHistoryEvent'
export JobQueueGraphQL from './jobQueue/types/JobQueue.graphql.js'
export Job, { MockJob } from './jobQueue/types/Job'
export JobGraphQL from './jobQueue/types/Job.graphql.js'
export JobFile, { MockJobFile } from './jobQueue/types/JobFile'

/* printer */
export PrinterGraphQL from './printer/types/Printer.graphql.js'

/* spool */
export * as PriorityEnum from './spool/types/PriorityEnum'
export * as TaskStatusEnum from './spool/types/TaskStatusEnum'
export * as ComponentTypeEnum from './config/types/ComponentTypeEnum'

export Task, { MockTask } from './spool/types/Task'
export TaskGraphQL from './spool/types/Task.graphql.js'
