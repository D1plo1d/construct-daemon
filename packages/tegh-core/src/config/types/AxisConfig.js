import { Record } from 'immutable'
import uuid from 'uuid/v4'

export const AxisConfigFactory = Record({
  id: null,
  name: null,
  feedrate: null,
})

const AxisConfig = props => (
  AxisConfigFactory({
    id: props.id || uuid(),
    ...props(),
  })
)

export default AxisConfig
