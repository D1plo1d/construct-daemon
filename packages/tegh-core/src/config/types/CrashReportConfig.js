import { Record, Map } from 'immutable'
import uuid from 'uuid'

export const CrashReportConfigFactory = Record({
  id: null,
  modelVersion: 0,
  directory: '/var/log/tegh',
  uploadCrashReportsToDevs: true,
  extendedConfig: Map(),
})

const CrashReportConfig = ({
  id = uuid.v4(),
  modelVersion = 0,
  ...props
} = {}) => (
  CrashReportConfigFactory({
    ...props,
    id,
    modelVersion,
    extendedConfig: Map(props.extendedConfig),
  })
)

export default CrashReportConfig
