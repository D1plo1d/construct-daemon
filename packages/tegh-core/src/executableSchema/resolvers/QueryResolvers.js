import getConfiguredDevices from '../../config/selectors/getConfiguredDevices'

const QueryResolvers = {
  Query: {
    /*
     * config
     */
    hosts: (source, args, { store }) => {
      const state = store.getState()

      if (args.hostID && args.hostID !== state.config.host.id) {
        return []
      }

      return [state.config.host]
    },
    materials: (source, args, { store }) => {
      const state = store.getState()

      if (args.materialID != null) {
        const materials = state.config.materials.find(c => (
          c.id === args.materialID
        ))
        if (materials == null) return []
        return [materials]
      }
      return state.config.materials
    },
    schemaForm: (source, args, { store }) => {
      const { collection, printerID, schemaFormKey } = args.input
      const state = store.getState()

      switch (collection) {
        case 'COMPONENT': {
          if (printerID !== state.config.printer.id) {
            throw new Error(`Printer ID: ${printerID} does not exist`)
          }
          const schemaForm = state.schemaForms.getIn(
            ['components', schemaFormKey],
          )
          return schemaForm
        }
        case 'MATERIAL': {
          const schemaForm = state.schemaForms.getIn(
            ['materials', schemaFormKey],
          )
          return schemaForm
        }
        default: {
          throw new Error(`Unsupported collection: ${collection}`)
        }
      }
    },
    /*
     * devices
     */
    devices: (_source, args, { store }) => {
      const state = store.getState()

      const connectedDevices = state.devices.byID
      // replace each configured device with it's connected equivalent if
      // it is connected.
      const configuredDevices = getConfiguredDevices(state.config)
        .map(device => (
          connectedDevices.find(d2 => d2.id === device.id) || device
        ))

      // remove duplicate devices
      const allDevices = configuredDevices.concat(connectedDevices.toList())
        .toOrderedSet()
        .toList()

      return allDevices
    },
    /*
     * jobQueue
     */
    jobQueue: (_source, args, { store }) => (
      store.getState()
    ),
    /*
     * printer
     */
    printers: (_source, args, { store }) => {
      const id = args.printerID
      const state = store.getState()
      if (id != null && id !== state.config.printer.id) {
        return []
      }
      return [state]
    },
  },
}

export default QueryResolvers
