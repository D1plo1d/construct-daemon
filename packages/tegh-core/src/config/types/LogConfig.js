import { Record, List, Map } from 'immutable'
import uuid from 'uuid/v4'

import { WARNING, ERROR, FATAL } from '../../log/types/logLevelEnum'

export const LogConfigFactory = Record({
  maxLength: 1000,
  stderr: List([WARNING, ERROR, FATAL]),
  extendedConfig: Map(),
})

const LogConfig = ({
  id = uuid(),
  ...props
} = {}) => (
  LogConfigFactory({
    ...props,
    id,
    extendedConfig: Map(props.extendedConfig),
  })
)

export default LogConfig
