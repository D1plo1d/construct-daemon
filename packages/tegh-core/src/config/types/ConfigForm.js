import { Record, Map } from 'immutable'
import t from 'tcomb-validation'

import LogConfig, { LogConfigStruct } from './LogConfig'
import CrashReportConfig, { CrashReportConfigStruct } from './CrashReportConfig'
import PrintFromLocalPathConfig, { PrintFromLocalPathConfigStruct } from './PrintFromLocalPathConfig'
import PluginConfig, { PluginConfigStruct } from './PluginConfig'
import MachineConfig, { MachineConfigStruct } from './MachineConfig'
import MaterialConfig, { MaterialConfigStruct } from './MaterialConfig'

export const configFormStructFields = {
  id: t.String,
  name: t.String,
  macros: t.dict(t.String, t.list(t.String)),
  log: LogConfigStruct,
  crashReports: CrashReportConfigStruct,
  printFromLocalPath: PrintFromLocalPathConfigStruct,
  materials: t.dict(t.String, MaterialConfigStruct),
  machine: MachineConfigStruct,
  plugins: t.dict(t.String, PluginConfigStruct),
}

export const ConfigFormStruct = t.struct(configFormStructFields)

export const ConfigFormRecordFactory = Record(
  Map(ConfigFormStruct.meta.props).map(() => null),
)

const mapOfRecords = (entries, recordFactory) => (
  Map(entries).map(props => recordFactory(props))
)

const ConfigForm = props => ConfigFormRecordFactory({
  ...props,
  log: LogConfig(props.log),
  crashReports: CrashReportConfig(props.crashReports),
  printFromLocalPath: PrintFromLocalPathConfig(props.printFromLocalPath),
  plugins: mapOfRecords(props.plugins, PluginConfig),
  machine: MachineConfig(props.machine),
  materials: mapOfRecords(props.materials, MaterialConfig),
})

export default ConfigForm
