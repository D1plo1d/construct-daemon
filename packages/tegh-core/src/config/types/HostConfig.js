import { Record, List, Map } from 'immutable'
import uuid from 'uuid/v4'

import CrashReportConfig from './CrashReportConfig'
import MaterialConfig from './MaterialConfig'
import LogConfig from './LogConfig'

export const HostConfigRecordFactory = Record({
  id: null,
  modelVersion: 0,
  name: null,
  crashReports: CrashReportConfig(),
  materials: List(),
  extendedConfig: Map(),
  log: LogConfig(),
})

const HostConfig = ({
  id = uuid(),
  modelVersion = 0,
  crashReports = {},
  materials = [],
  log = {},
  ...props
} = {}) => (
  HostConfigRecordFactory({
    ...props,
    id,
    modelVersion,
    crashReports: CrashReportConfig(crashReports),
    materials: materials.map(MaterialConfig),
    log: LogConfig(log),
    extendedConfig: Map(props.extendedConfig),
  })
)

export default HostConfig
