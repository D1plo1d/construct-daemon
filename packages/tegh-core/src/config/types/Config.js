import { Record, List } from 'immutable'
import uuid from 'uuid/v4'

import HostConfig from './HostConfig'
import PrinterConfig from './PrinterConfig'
import MaterialConfig from './MaterialConfig'
import developmentConfig from '../../../../../development.config'

export const ConfigRecordFactory = Record({
  id: null,
  modelVersion: 0,
  host: null,
  printer: null,
  materials: List(),
})

const Config = ({
  id = uuid(),
  modelVersion = 0,
  host = {},
  printer = {},
  materials = [],
  ...props
} = {}) => (
  ConfigRecordFactory({
    id,
    modelVersion,
    host: HostConfig(host),
    printer: PrinterConfig(printer),
    materials: List(materials).map(material => MaterialConfig(material)),
    ...props,
  })
)

export const MockConfig = ({
  host = {},
  printer = {},
  ...props
} = {}) => (
  Config({
    ...developmentConfig,
    host: {
      ...developmentConfig.host,
      ...host,
    },
    printer: {
      ...developmentConfig.printer,
      ...printer,
    },
    ...props,
  })
)

export default Config
