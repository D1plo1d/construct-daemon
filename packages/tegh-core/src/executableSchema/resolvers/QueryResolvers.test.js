import { Record, Map } from 'immutable'
import typeDefs from '@tegh/schema'

import deterministicTestSetup from '../../util/testing/deterministicTestSetup'
import snapshotTestResolvers from '../../util/testing/snapshotTestResolvers'

import { MockConfig } from '../../config/types/Config'

import QueryResolvers from './QueryResolvers'

describe('QueryResolvers', () => {
  it('presents a consistent API', () => {
    deterministicTestSetup()

    const state = Record({
      config: MockConfig(),
      schemaForms: Record({
        components: Map({
          TOOLHEAD: Record({
            id: 'TOOLHEAD',
          })(),
        }),
      })(),
      devices: Record({
        byID: Map({}),
      })(),
    })()

    snapshotTestResolvers({
      typeDefs,
      typeName: 'Query',
      resolvers: QueryResolvers,
      fieldArgs: {
        args: {},
        children: {
          schemaForm: {
            args: {
              input: {
                collection: ['__ENUM', 'COMPONENT'],
                printerID: state.config.printer.id,
                schemaFormKey: 'TOOLHEAD',
              },
            },
          },
        },
      },
      rootValue: {},
      contextValue: {
        store: {
          getState: () => state,
        },
      },
    })
  })
})
