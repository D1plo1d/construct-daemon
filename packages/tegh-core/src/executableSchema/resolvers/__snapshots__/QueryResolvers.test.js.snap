// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryResolvers presents a consistent API 1`] = `
Object {
  "_testQuery": Object {
    "_testQuery": Object {},
    "devices": Array [
      Object {
        "id": "/dev/serial/by-id/usb-Arduino__www.arduino.cc__Arduino_Mega_2560_749373037363518101E2-if00",
      },
    ],
    "hostConfigs": Array [
      Object {
        "id": "pzxcvkkwn",
      },
    ],
    "jobQueue": Object {
      "id": "pzxcvkkwn",
    },
    "materials": Array [
      Object {
        "id": "dkdow2",
      },
      Object {
        "id": "bw021",
      },
    ],
    "printerConfigs": Array [
      Object {
        "id": "aaxcvxcvcxv-bvb-csdf234231",
      },
    ],
    "printers": Array [
      Object {
        "id": "aaxcvxcvcxv-bvb-csdf234231",
      },
    ],
    "schemaForm": Object {
      "id": "TOOLHEAD",
    },
  },
}
`;
